pkg/token/serviceprincipaltoken.go

```diff
@@ -69,91 +73,177 @@ func (p *servicePrincipalToken) Token() (adal.Token, error) {
+	}
+	accessToken, expiresOn, err := pop.AcquirePoPTokenConfidential(
+		context,
+		p.popClaims,
+		scopes,
+		cred,
+		p.cloud.ActiveDirectoryAuthorityHost,
+		p.clientID,
+		p.tenantID,
+		options,
+	)
+	if err != nil {
+		return "", -1, fmt.Errorf("failed to create service principal PoP token using secret: %w", err)
+	}
+
+	return accessToken, expiresOn, nil
+}
+
+// requests a token using the configured client ID/certificate and returns a PoP token if PoP claims are provided,
+// otherwise returns a regular bearer token
+func (p *servicePrincipalToken) getTokenWithClientCert(context context.Context, scopes []string, options *azcore.ClientOptions) (string, int64, error) {
+	clientOptions := &azidentity.ClientCertificateCredentialOptions{
+		ClientOptions: azcore.ClientOptions{
+			Cloud: p.cloud,
+		},
+		SendCertificateChain: true,
+	}
+	if options != nil {
+		clientOptions.ClientOptions = *options
+	}
+	certData, err := os.ReadFile(p.clientCert)
+	if err != nil {
+		return "", -1, fmt.Errorf("failed to read the certificate file (%s): %w", p.clientCert, err)
+	}
+
+	// Get the certificate and private key from pfx file
+	cert, rsaPrivateKey, err := decodePkcs12(certData, p.clientCertPassword)
+	if err != nil {
+		return "", -1, fmt.Errorf("failed to decode pkcs12 certificate while creating spt: %w", err)
+	}
+
+	certArray := []*x509.Certificate{cert}
+	if len(p.popClaims) > 0 {
+		// if PoP token support is enabled, use the PoP token flow to request the token
+		return p.getPoPTokenWithClientCert(context, scopes, certArray, rsaPrivateKey, options)
+	}
+
+	cred, err := azidentity.NewClientCertificateCredential(
+		p.tenantID,
+		p.clientID,
+		certArray,
+		rsaPrivateKey,
+		clientOptions,
+	)
+	if err != nil {
+		return "", -1, fmt.Errorf("unable to create credential. Received: %v", err)
+	}
+	spnAccessToken, err := cred.GetToken(context, policy.TokenRequestOptions{Scopes: scopes})
+	if err != nil {
+		return "", -1, fmt.Errorf("failed to create service principal token using cert: %s", err)
+	}
+
+	return spnAccessToken.Token, spnAccessToken.ExpiresOn.Unix(), nil
+}
+
+// requests a PoP token using the given client ID/certificate and returns it
+func (p *servicePrincipalToken) getPoPTokenWithClientCert(
+	context context.Context,
+	scopes []string,
+	certArray []*x509.Certificate,
+	rsaPrivateKey *rsa.PrivateKey,
+	options *azcore.ClientOptions,
+) (string, int64, error) {
+	cred, err := confidential.NewCredFromCert(certArray, rsaPrivateKey)
+	if err != nil {
+		return "", -1, fmt.Errorf("unable to create credential from certificate. Received: %w", err)
+	}
+
+	accessToken, expiresOn, err := pop.AcquirePoPTokenConfidential(
+		context,
+		p.popClaims,
+		scopes,
+		cred,
+		p.cloud.ActiveDirectoryAuthorityHost,
+		p.clientID,
+		p.tenantID,
+		options,
+	)
+	if err != nil {
+		return "", -1, fmt.Errorf("failed to create service principal PoP token using certificate: %w", err)
+	}
+
+	return accessToken, expiresOn, nil
+}
+
```

Comment:
@weinong should we split this into 2 files `serviceprincipaltokensecret.go` and `serviceprincipaltokencert.go`, this is starting to be a lot of lines ?

Reply:
sounds like a good idea!

Reply:
Split this into 3 files and added additional UTs to increase coverage